### PR TITLE
Change command saving to the buffer. Add WiFi connection check.

### PR DESCRIPTION
**Pull request by vadmatura**

### CODE REVIEW

File: arduino_currency.ino [hunk 1]
```diff
@@ -4,7 +4,8 @@
 
 #define SERIAL_BIT_RATE 9600
 typedef void (*DataCompleteFunc)(char* dataStr);
-typedef void (*CommandsCompleteFunc)(uint8_t commandNum, bool error);
+typedef void (*CommandResultFunc)(uint8_t commandNum, const char** commandArray, char* commandResponce);
```
Comment (iostaf) [inline arduino_currency.ino]: Typo in `commandResponce`, should be _Response_

File: arduino_currency.ino [hunk 2]
```diff
@@ -87,10 +89,11 @@ public:
           m_dataCounter--;
           if (m_dataCounter == 0) {
             Serial.print("\r\n----DATA----");
-            changeWorkMode(WorkMode::dataHead);
+            changeWorkMode(WorkMode::dataNextHead);
           }
+          m_msecTimeWork = millis();
         } else {
-          addToShiftCommandBuffer(responceChar);
+          addToCommandBuffer(responceChar);
```
Comment (iostaf) [inline arduino_currency.ino]: Again `responce` typo :)

File: arduino_currency.ino [hunk 3]
```diff
@@ -101,7 +104,7 @@ public:
             changeWorkMode(WorkMode::dataHead);
             break;
           case AtResult::ok_dataHead:
-            headPos = strstr(m_shiftCommandBuffer, "+IPD,");
+            headPos = strstr(m_commandBuffer, "+IPD,");
             if ( headPos != NULL) {
               dataLen = String(headPos + 5);
```
Comment (iostaf) [inline arduino_currency.ino]: Avoid magic number, if it a size of `"+IPD,"` it's better to get string size

File: arduino_currency.ino [hunk 4]
```diff
@@ -123,62 +126,70 @@ public:
           case AtResult::none:
             break;
           default:
-            Serial.println("\r\n----END----");
+            Serial.println("\r\n----UNCKNOWN RESULT----");
```
Comment (iostaf) [inline arduino_currency.ino]: Typo in `UNCKNOWN`, please use _UNKNOWN_ instead

File: arduino_currency.ino [hunk 5]
```diff
@@ -123,62 +126,70 @@ public:
...
-  void addToShiftCommandBuffer(char c) {
-    memcpy(m_shiftCommandBuffer, m_shiftCommandBuffer+1, CM_SHIFT_COMMAND_BUFFER_LAST);
-    m_shiftCommandBuffer[CM_SHIFT_COMMAND_BUFFER_LAST] = c;
+  void addToCommandBuffer(char c) {
+    if (m_commandBufferPos < m_commandBufferLast) {
+      *m_commandBufferPos = c;
+      m_commandBufferPos++;
+      *m_commandBufferPos = '\0';
+    } else {
+      finishWork(true);
```
Comment (iostaf) [inline arduino_currency.ino]: #Question: So, if I get it right, we just finish work in case of buffer overflow?

File: arduino_currency.ino [hunk 6]
```diff
@@ -123,62 +126,70 @@ public:
...
+    m_commandBufferPos = m_commandBuffer;
+    *m_commandBufferPos = '\0';
   }
 
   AtResult checkResult() {
     switch(m_workMode) {
     case WorkMode::dataHead:
-      if (strstr(m_shiftCommandBuffer, ":") != NULL) {
+    case WorkMode::dataNextHead:
+      if (strstr(m_commandBuffer, ":") != NULL) {
```
Comment (iostaf) [inline arduino_currency.ino]: #Question: What is a structure of the m_commandBuffer, is it contains only one command or several ones in the string?

File: arduino_currency.ino [hunk 7]
```diff
@@ -123,62 +126,70 @@ public:
...
     }
     return AtResult::none;
   }
 
   void call() {
-    if (m_numCommandForCall < m_countForCall) {
+    if (m_countForCall > 0) {
       Serial.println("\r\n--------");
       esp8266.write(m_commandArray[m_numCommandForCall]);
       esp8266.write("\r\n");
```
Comment (iostaf) [inline arduino_currency.ino]: I've seen `\r\n` sequence a few times already, maybe it worth to extract it as a constant?

File: arduino_currency.ino [hunk 8]
```diff
@@ -191,25 +202,23 @@ protected:
...
     m_numCommandForCall++;
+    m_countForCall--;
     m_callRepeatCount = 0;
     call();
   }
 
   void callRepeat() {
     m_callRepeatCount++;
-    if (m_callRepeatCount > 3) {
+    if (m_callRepeatCount > 2) {
```
Comment (iostaf) [inline arduino_currency.ino]: Please extract magic number

File: arduino_currency.ino [hunk 9]
```diff
@@ -274,28 +274,34 @@ private:
...
-                          "AT", 
-                          "AT", 
-                          "AT", 
-                          "AT", 
-                          "AT", 
-                          "AT", 
-                          "AT", 
-                          "AT", 
-                          "AT"};
+//"AT+RST"
```
Comment (yurii-krupa) [inline arduino_currency.ino]: Need more meaningful comments, please provide )

File: arduino_currency.ino [hunk 10]
```diff
@@ -252,40 +266,118 @@ private:
...
+bool isConnectedToTCP = false;
+CommandManager cm(&dataCompleteFunc, &commandResultFunc, &commandsCompleteFunc);
+
+#define MAX_FLOAT_STR_LENGTH 12
+
+float getFloatValueByKey(char* dataStr, const char* key) {
+  float ret = 0.0;
+  char* cStart = strstr(dataStr, key);
+  if ( cStart != NULL) {
+    cStart += strlen(key) + 3; // ":" - 3 symbols
```
Comment (iostaf) [inline arduino_currency.ino]: I do not understand this one. Could you please jot down a comment to explain it to me and probably future self.

File: arduino_currency.ino [hunk 11]
```diff
@@ -252,40 +266,118 @@ private:
...
+      *cEnd = '\"';
+    }
+  }
+  return ret;
+}
 
 void dataCompleteFunc(char* dataStr) {
-  Serial.println("\r\n\r\nFIND DATA!!!");
-  Serial.println(dataStr);
+  /*Serial.print("\r\nFIND DATA: ");
```
Comment (iostaf) [inline arduino_currency.ino]: Do you really want this commented dead code to stay here?

File: arduino_currency.ino [hunk 12]
```diff
@@ -252,40 +266,118 @@ private:
...
+    Serial.println(buy);
+  }
 }
 
-void commandsCompleteFunc(uint8_t commandNum, bool error) {
-  Serial.println("\r\n\r\nEND COMMANDS!!!");
-  Serial.println(commandNum);
-  Serial.println(error?"error":"ok");
+void commandResultFunc(uint8_t commandNum, const char** commandArray, char* commandResponce) {
+  /*Serial.print("\r\nCOMMAND: ");
```
Comment (iostaf) [inline arduino_currency.ino]: Do you really want this commented dead code to stay here?

File: arduino_currency.ino [hunk 13]
```diff
@@ -252,40 +266,118 @@ private:
...
+        Serial.println("ConnectedToTCP");
+    } else {
+      isConnectedToTCP = false;
+    }
+  }
 }
 
-CommandManager cm(commands, &dataCompleteFunc, &commandsCompleteFunc);
+void commandsCompleteFunc(uint8_t commandNum, const char** commandArray, bool error) {
+  /*Serial.print("\r\n\r\nEND COMMANDS: ");
```
Comment (iostaf) [inline arduino_currency.ino]: Do you really want this commented dead code to stay here?

File: arduino_currency.ino [hunk 14]
```diff
@@ -252,40 +266,118 @@ private:
...
+      } else {
+        if (isConnectedToWiFi) {
+          cm.call(commandsGetData);
+        } else {
+          cm.call(commandsConnectToWifi);
+        }
+      }
+    }
+  } else {
+    //cm.call(commandsStart);
```
Comment (iostaf) [inline arduino_currency.ino]: Wat???

File: arduino_currency.ino [hunk 15]
```diff
@@ -4,8 +4,8 @@
 
 #define SERIAL_BIT_RATE 9600
```
Comment (yurii-krupa) [inline arduino_currency.ino]: Move defining SERIAL_BIT_RATE closer to it's concrete place of using.
I've seen it somewhere after line 380.

File: arduino_currency.ino [hunk 16]
```diff
@@ -328,36 +334,48 @@ void dataCompleteFunc(char* dataStr) {
   }
```
Comment (yurii-krupa) [inline arduino_currency.ino]: Line 312 - thinking that would be better to replace "+ 3" with some const or it have no sense?

File: arduino_currency.ino [hunk 17]
```diff
@@ -4,8 +4,8 @@
 
 #define SERIAL_BIT_RATE 9600
 typedef void (*DataCompleteFunc)(char* dataStr);
-typedef void (*CommandResultFunc)(uint8_t commandNum, char* commandStr);
-typedef void (*CommandsCompleteFunc)(uint8_t commandNum, bool error);
+typedef void (*CommandResultFunc)(uint8_t commandNum, const char** commandArray, char* commandResponce);
+typedef void (*CommandsCompleteFunc)(uint8_t commandNum, const char** commandArray, bool error);
```
Comment (yurii-krupa) [inline arduino_currency.ino]: Comman**d**CompleteFunc or Command**S**CompleteFucn :)

File: arduino_currency.ino [hunk 18]
```diff
@@ -4,8 +4,8 @@
...
 
 SoftwareSerial esp8266(2, 3);
```
Comment (yurii-krupa) [inline arduino_currency.ino]: We need think about places of var announcing it complicates navigation.
Or I'm making up comments just for comments ))) For me it looks little messy

File: arduino_currency.ino (hunk 9 above)
Comment (vadmatura) [reply arduino_currency.ino]: done

File: arduino_currency.ino (hunk 1 above)
Comment (vadmatura) [reply arduino_currency.ino]: Done

File: arduino_currency.ino (hunk 2 above)
Comment (vadmatura) [reply arduino_currency.ino]: Done

File: arduino_currency.ino (hunk 3 above)
Comment (vadmatura) [reply arduino_currency.ino]: Done

File: arduino_currency.ino (hunk 4 above)
Comment (vadmatura) [reply arduino_currency.ino]: Done

File: arduino_currency.ino (hunk 11 above)
Comment (vadmatura) [reply arduino_currency.ino]: Add #define DEBUG_SERIAL

File: arduino_currency.ino (hunk 12 above)
Comment (vadmatura) [reply arduino_currency.ino]: Add DEBUG_SERIAL

File: arduino_currency.ino (hunk 13 above)
Comment (vadmatura) [reply arduino_currency.ino]: Add DEBUG_SERIAL

File: arduino_currency.ino (hunk 14 above)
Comment (vadmatura) [reply arduino_currency.ino]: Done

File: arduino_currency.ino (hunk 15 above)
Comment (vadmatura) [reply arduino_currency.ino]: Done

File: arduino_currency.ino (hunk 17 above)
Comment (vadmatura) [reply arduino_currency.ino]: Call CommandSCompleteFucn - after all commands are finished.

**Pull request merged**